packages/controllers/src/permissions/PermissionController.test.ts

```diff
@@ -0,0 +1,4014 @@
+      const { messenger } = options;
+      const origin = 'metamask.io';
+      const controller = getDefaultPermissionController(options);
+
+      const callActionSpy = jest
+        .spyOn(messenger, 'call')
+        .mockImplementationOnce((async (...args: any) => {
+          const [, { requestData }] = args;
+          return {
+            metadata: { ...requestData.metadata },
+            permissions: {
+              [PermissionNames.wallet_getSecretArray]: {
+                parentCapability: PermissionNames.wallet_getSecretArray,
+              },
+              // parentCapability value does not match key
+              [PermissionNames.wallet_getSecretObject]: {
+                parentCapability: PermissionNames.wallet_getSecretArray,
+              },
+            },
+          };
+        }) as any);
+
+      await expect(
+        async () =>
+          await controller.requestPermissions(
+            { origin },
+            {
+              [PermissionNames.wallet_getSecretArray]: {
+                parentCapability: PermissionNames.wallet_getSecretArray,
+              },
+            },
+          ),
+      ).rejects.toThrow(
+        errors.invalidParams({
+          message: `Invalid approved permissions request: Permissions request for origin "${origin}" contains invalid requested permission(s).`,
+          data: {
+            origin,
+            requestedPermissions: {
+              [PermissionNames.wallet_getSecretArray]: {
+                [PermissionNames.wallet_getSecretArray]: {
+                  parentCapability: PermissionNames.wallet_getSecretArray,
+                },
+                [PermissionNames.wallet_getSecretObject]: {
+                  parentCapability: PermissionNames.wallet_getSecretArray,
+                },
+              },
+            },
+          },
+        }),
+      );
+
+      expect(callActionSpy).toHaveBeenCalledTimes(1);
+      expect(callActionSpy).toHaveBeenCalledWith(
+        'ApprovalController:addRequest',
+        {
+          id: expect.any(String),
+          origin,
+          requestData: {
+            metadata: { id: expect.any(String), origin },
+            permissions: {
+              [PermissionNames.wallet_getSecretArray]: {
+                parentCapability: PermissionNames.wallet_getSecretArray,
+              },
+            },
+          },
+          type: MethodNames.requestPermissions,
+        },
+        true,
+      );
+    });
+  });
+
+  describe('acceptPermissionsRequest', () => {
+    it('accepts a permissions request', async () => {
+      const options = getPermissionControllerOptions();
+      const { messenger } = options;
+      const origin = 'metamask.io';
+      const id = 'foobar';
+
+      const callActionSpy = jest
+        .spyOn(messenger, 'call')
+        .mockImplementationOnce((async (..._args: any) => true) as any)
+        .mockImplementationOnce((async (..._args: any) => undefined) as any);
```

Comment:
These casts aren't needed, and these actions are both synchronous

```suggestion
        .mockImplementationOnce((..._args: any) => true)
        .mockImplementationOnce((..._args: any) => undefined);
```

Reply:
I see the same mistakes are made in a few subsequent tests as well

Reply:
Fixed these and a bunch of others in 98d12cb